File: Data Structures & Algorithms/trapping-rain-water/submission-8.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # Can be solved monotonous stack-> PGE+NSE
        n=len(height)
        s=[]
        total=0
        for i in range(n):
            while s and height[i] > height[s[-1]] :
                floor=height[s.pop()]
                if not s : break
                left=s[-1]
                width=i-left-1
                amt=width*(min(height[left],height[i])-floor)
                total+=amt
            s.append(i)
        
        return total

File: Data Structures & Algorithms/climbing-stairs/submission-8.cpp
class Solution {
public:
int func(int n,vector<int> &dp){
    if(n<=1) return 1;
    if(dp[n]!=-1) return dp[n];
    dp[n]=func(n-2,dp)+func(n-1,dp);
    return dp[n];
 }
    int climbStairs(int n) {
        vector<int> dp(n+1,-1);
        return func(n,dp);
        
        
    }
};
